### PR TITLE
[SPARK-22530][PYTHON][SQL] Adding Arrow support for ArrayType

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

This change adds `ArrayType` support for working with Arrow in pyspark when creating a DataFrame, calling `toPandas()`, and using vectorized `pandas_udf`.

## How was this patch tested?

Added new Python unit tests using Array data.

### CODE REVIEW

**Comment by BryanCutler** [comment]
ping @ueshin @HyukjinKwon

Unfortunately, there was a bug in the Arrow 0.8.0 release on the Java side https://issues.apache.org/jira/browse/ARROW-1948 that caused a problem here.  I was able to find a workaround, but it required me to make a change to the `ArrowVectorAccessor` class.  I'm not sure if this is something you would be ok putting in, or if you would prefer to wait until the next minor release to add the ArrayType support.

The issue was that the Arrow spec states that if the validity buffer is empty, then that means that all the values are non-null.  In Arrow 0.8.0, the C++/Python side started sending buffers this way, and the Arrow ListVector was not handling it properly, thinking instead that there were no valid values.  

The workaround I added here looks if the ListVector has a value count of > 0 and has an empty validity buffer.  This means that all the values are non-null and it will allocate a new validity buffer with all bits set.

For Arrow with non-udfs (toPandas and createDataFrame) this only needs to be done once, but for udfs each batch read will load new buffers into the arrow VectorSchemaRoot, so it needs to be checked after each read.  The simplest place to put the workaround to cover these cases was to allow `ArrowVectorAccessor.isNullAt(int rowId)` to be overridden.  Let me know what you guys think, thanks!

**Comment by ueshin** [comment]
How about simply returning `false` from `ArrowVectorAccessor.isNullAt(int rowId)` when `accessor.getValueCount() > 0 && accessor.getValidityBuffer().capacity() == 0` without modifying the buffer?

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by BryanCutler** [comment]
> How about simply returning false from ArrowVectorAccessor.isNullAt(int rowId) when accessor.getValueCount() > 0 && accessor.getValidityBuffer().capacity() == 0

Good idea @ueshin , I think this should be fine as we are only querying the validity buffer in the call to `isNullAt`.  I'll give it a try!

**Comment by BryanCutler** [comment]
The new workaround seems to be fine and I also added another test with array null values to test that along with all non-null values.

**Review by HyukjinKwon — APPROVED** [review]
LGTM

File: python/pyspark/sql/tests.py [hunk 1]
```diff
@@ -3357,6 +3357,27 @@ def test_schema_conversion_roundtrip(self):
...
+        for r in range(len(expected)):
+            for e in range(len(expected[r])):
+                self.assertTrue(expected[r][e] == result[r][e])
+
+    def test_toPandas_with_array_type(self):
+        expected = [([1, 2], [u"x", u"y"]), ([3, 4], [u"y", u"z"])]
+        array_schema = StructType([StructField("a", ArrayType(IntegerType())),
+                                   StructField("b", ArrayType(StringType()))])
+        df = self.spark.createDataFrame(expected, schema=array_schema)
+        pdf_arrow = df.toPandas()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: Should we maybe add some tests to check if they are the same when arrow is enabled/disabled for sure?

Comment (BryanCutler) [reply python/pyspark/sql/tests.py]: Sure, I can add that

Comment (BryanCutler) [reply python/pyspark/sql/tests.py]: ok done

**Comment by BryanCutler** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by BryanCutler** [comment]
Thanks @HyukjinKwon and @ueshin !

**Pull request closed without merging**